### DISCUSSION TITLE
Default implementations

### DISCUSSION THREAD

**Discussion by Pyrdacor** [dotnet/csharplang#8222] [Language Ideas]
The idea is to allow a short language syntax to auto-implement default behaviors by the compiler. This proposal mainly focuses on functions. What comes to mind are things like `GetHashCode` or `Clone`. For me the latter is quite useful as currently there is no good way to clone objects (especially deep cloning).

With records we now have classes with some default behavior auto-implemented already. So something similar was done before.

However the new keyword allows more flexible approaches and more control. And it could in theory be used for any type of object.

```cs
class MyClass : ICloneable
{
    public int MyProp { get; set; }

    default object Clone();
    default int GetHashCode();
}
```

In this case the compiler would generate something like this:

```cs
class MyClass : ICloneable
{
    public int MyProp { get; set; }

    public object Clone()
    {
        return new MyClass
        {
            MyProp = MyProp
        };
    }

    public int GetHashCode()
    {
        return HashCode.Combine(MyProp);
    }
}
```

It might also work with patterns like `IDisposable` or `INotifyPropertyChanged`. So basically for well-known patterns which have a default behavior most of the time. This saves a lot of time and boilerplate code. Especially cloning and hash code generation would be enough to justify such language feature imo.

So the accepted functions might be limited by the compiler and it can generate an error if the function has no default behavior. This can be extended later if needed.

This idea can of course be extended or adjusted. My goal is to allow cloning easier and save boilerplate code for classes.

**Comment by HaloFour** [comment]
Feels like something better suited to a source generator, one that can be configurable and apply policies to the emitted implementation of those methods.

**Reply by FaustVX** [reply]
Yes, SG is the way to go here

**Reply by Pyrdacor** [reply]
But it should be provided by the runtime then I think. And how would you apply this to the class? Based on some attribute like `[AutoClone]`?

**Reply by Pyrdacor** [reply]
The main question I see is how to tell the compiler or SG that the default behavior should be used. In the end it isn't that important who adds the code on the fly. It is more about how to express the intention of using the default behavior.

**Reply by FaustVX** [reply]
You could add `[AutoClone, AutoHashCode]` on classes, then the SG would generate the `Clone` and `GetHashCode` method, then it's compiled into regular classes

**Reply by Pyrdacor** [reply]
And who provides the SG and attributes? I think basic behaviors should be provided by the compiler, the runtime or some other core component. In the end interfaces like `ICloneable` or `INotifyPropertyChanged` were designed with some specific workflow in mind. I think it makes sense to provide the attributes and SGs alongside with them then.

**Reply by FaustVX** [reply]
> And who provides the SG and attributes?

Anyone who want to create this NuGet package

**Reply by Pyrdacor** [reply]
So I guess nobody. Up to today there is only one nuget which does it and it's not maintained and only works partially. So I guess we'll stick to JSON Serdes or cloning via reflection. If this has to be provided by 3rd party nuget packages, there won't be an unified approach to this. Maybe cloning is not that important nowadays. Still feels a bit sad that C# has no satisfying solution for such fundamental feature.

The benefit to provide it by the core is that it can automatically be adjusted if the language or framework changes. For example if special collections are added. No 3rd party nuget will maintain all the changes ever. So this is no real solution.

**Reply by TahirAhmadov** [reply]
> So I guess nobody.

That means it will be you. :)

**Reply by iam3yal** [reply]
@Pyrdacor 

> this is no real solution.

With all due respect we see this a lot here, the moment someone mentions SG or 3rd party packages some guys here go but this is a "workaround", "not a real solution" or whatever but if you REALLY need it then YOU should definitely create it and maintain it yourself, don't you think? because the mechanism exists but some of you just don't want to do it and expect other devs to do it for you. 

So I'll say this if it's not a real solution you probably don't have a real problem.

**Reply by Pyrdacor** [reply]
I didn't say that I couldn't implement it myself. But if the implementation is only inside my own 3rd party nuget package, this is not really useful to have a general deep cloning mechanism in C#.

**Reply by Pyrdacor** [reply]
> > So I guess nobody.
> 
> That means it will be you. :)

Wow didn't expect insults in a place like this, but ok.

**Reply by TahirAhmadov** [reply]
> I didn't say that I couldn't implement it myself. But if the implementation is only inside my own 3rd party nuget package, this is not really useful to have a general deep cloning mechanism in C#.

Deep cloning is not really a well defined set of requirements. I struggle with it in my code, and that's just one code base (admittedly, with many many solutions). In some cases, there are recursive references, in others, some classes need to be excluded from being cloned due to business requirements, etc. It's definitely not something which can be added to the language for everybody's usage.

> Wow didn't expect insults in a place like this, but ok.

You misunderstood my comment. I didn't mean to say "you are a nobody", what I meant was "since nobody else is willing to do it, then it will have to be you". Basically the same thing that @iam3yal said:

> if you REALLY need it then YOU should definitely create it and maintain it yourself, don't you think? because the mechanism exists but some of you just don't want to do it and expect other devs to do it for you.

Believe me, neither me nor anybody else here are trying to insult you.

**Reply by Pyrdacor** [reply]
> @Pyrdacor
> 
> > this is no real solution.
> 
> With all due respect we see this a lot here, the moment someone mentions SG or 3rd party packages some guys here go but this is a "workaround", "not a real solution" or whatever but if you REALLY need it then YOU should definitely create it and maintain it yourself, don't you think? because the mechanism exists but some of you just don't want to do it and expect other devs to do it for you.
> 
> So I'll say this if it's not a real solution you probably don't have a real problem.

Maybe a few words about that. It's not about laziness in those cases. It's about trust in the expertise. I could also repair my car myself and I do it regularly if it's a small thing. But for most things I go to my nearby car repair. I could also ask a friend who says he can repair cars, but in general I still go to the experts for that.

3rd party libraries often bring this feeling with them, that the author didn't do it the best way. Having features inside the runtime or compiler feels safe and everybody thinks that it is most likely implemented in the best possible way. Moreover it is much more reviewed and discussed. And there are many people behind it so that maintenance is guaranteed.

In the end everybody in these discussions wants to improve something. I don't think that most people just are lazy and want others to do their work.
